### ISSUE TITLE
Flickering block in ATV Offroad Fury Pro

### ISSUE THREAD

**Issue by daniel229** [hrydgard/ppsspp#8481]
Also in the demo.
demo download
https://drive.google.com/file/d/0BzGZGDfFE68zUG9KT2YxV1J1azg/view?usp=sharing

savestate for demo
https://drive.google.com/file/d/0BzGZGDfFE68zNHZ3eVluOEtTZ3c/view?usp=sharing

![12](https://cloud.githubusercontent.com/assets/3481559/12528063/2264756c-c1c6-11e5-95d5-ac5981e78162.png)

GE debugger
![13](https://cloud.githubusercontent.com/assets/3481559/12528062/2263e746-c1c6-11e5-8733-07849a605fe7.png)
https://gist.github.com/daniel229/170dce2a1478697b6c8b

**Comment by unknownbrackets** [comment]
Pretty sure this is an issue with mipmapping.

Tex filtering min: linear, mipmap linear, mag: linear
Tex level/bias	slope + bias: 0.000000
Tex lod slope	0.199997

Because it's grayscale, and using add src * dst + src * dst, the lower mipmap levels are darker.  Without the slope bias, we sometimes use the wrong miplevel.

-[Unknown]

**Comment by unknownbrackets** [comment]
It's currently flickering in softgpu - and using slope.  So this probably indicates that slope is more complicated after all (maybe it takes into account w or something?)

Here's a video, just so we can be sure it's not a game bug.  It's clearly not flickering:
https://youtu.be/QRkLhrftxZc?t=3m4s

The texture is 0x040e5b40.  It's pretty clear that from far away, it should be using the higher (smaller) mip levels, and closer it should be using the nearer ones, which fixes it.  Otherwise, it looks far too grainy and flickers (in GLES and in softgpu.)

-[Unknown]

**Comment by Gamemulatorer** [comment]
So this is still can reproduce
https://github.com/hrydgard/ppsspp/issues/14984#issuecomment-938213957

**Comment by Gamemulatorer** [comment]
Still flickering.

https://user-images.githubusercontent.com/37603562/190122057-de5d1450-dbfd-4ef2-b267-2084dc811b00.mp4

**Comment by Gamemulatorer** [comment]
I can also reproduce this issue on software mode.
![Screenshot_2022-09-14-17-53-48-79](https://user-images.githubusercontent.com/37603562/190123105-10353f9b-a2fe-4402-9b1b-0f913708a4f1.png)
[UCES00786.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/9565136/UCES00786.ppdmp.zip)

**Comment by Gamemulatorer** [comment]
ppsspp v1.13.2 driver is not pink 🤔

Software 
![Screenshot_2022-09-19-19-59-03-19](https://user-images.githubusercontent.com/37603562/191012630-ac612398-3b6d-4575-aa20-75f2ceeb0c46.png)

Hardware 
![Screenshot_2022-09-19-19-59-27-67](https://user-images.githubusercontent.com/37603562/191012667-f355da70-41eb-4745-ab6f-7ee2412291ef.png)

**Comment by Gamemulatorer** [comment]
Now it's black using the latest build.
![Screenshot_2022-09-19-22-51-27-78](https://user-images.githubusercontent.com/37603562/191047417-68f03f3f-6822-4c6f-92e2-1cb18515ba8f.png)
[UCES00786.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/9600037/UCES00786.ppdmp.zip)

**Comment by Gamemulatorer** [comment]
Software is correct now using the latest build https://github.com/hrydgard/ppsspp/commit/913ba9de88b2dbd07d045be1fc1c5ff6cd62212a
![Screenshot_2022-09-20-08-40-19-51](https://user-images.githubusercontent.com/37603562/191142893-564d5023-5be8-49c5-b2a3-057be5c90542.png)

**Comment by Gamemulatorer** [comment]
The driver still black using the recently build.

OPENGL
![Screenshot_20221211_012305_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/206867414-93d97e07-eef0-4727-b325-1e3abf35466d.jpg)

VULKAN
![Screenshot_20221211_012322_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/206867425-36091ef0-82e0-4c9c-92c2-f47609971ee7.jpg)

**Comment by hrydgard** [comment]
Alright, what's actually going on (barring all the z fighting issue and stuff above, which I do not see):

It draws the level in chunks, first the base layer which is a low resolution texture of the whole level, then it applies a repeating detail texture on top, which brightens things - and this is a problem because it only applies it on blocks that are up to a certain distance from the camera, it seems. So this brightening is visibly flickering into existence.

(video placeholder for later).

These detail textures are drawn on top using the following rather odd blend setup:

![image](https://user-images.githubusercontent.com/130929/219011661-d4bb01b2-b8d2-47f3-9492-fa6319b9f8bd.png)

We can ignore alpha here since it's not used as a color factor. But the colors themselves are, and effectively this means:

srcCol * dstCol + dstCol * srcCol, or 2*srcCol*dstCol. If the average of the detail texture was a 0.5 gray, this would work fine and wouldn't look off against the distance where the blending doesn't happen, but, well, it's not, at least not the top mipmaps, which look like this: 

![image](https://user-images.githubusercontent.com/130929/219012085-3db60091-ff9a-43eb-9116-1bd3d06ad20c.png)

The fourth mipmap, however, at 8x8, is full with color 128. So if the game uses trilinear filtering, and makes sure that at the distance where the fourth mip kicks in is less than the distance at which is culls these draws, then the brightness mismatch would at least just fade in instead of pop. And this seems to be what's happening - in the software renderer, the ground is unnaturally brighter close to the camera using this effect. It doesn't look bad actually, might even be intended.

Fixing this in higher resolutions in a way that looks good is non-trivial - if we just push the mip bias to match, we'll pick very low resolution mipmaps and the ground will look blurry. 

The best fix here is likely to be a texture replacement. In the meantime, playing on a fast PC using the software renderer will be the best experience here.

**Comment by unknownbrackets** [comment]
I'll note that we do have a slightly inaccurate slope implementation in software that is not implemented in hardware rendering.  See #16276 (d18fc5c85dfa67b428e4fbfb2a0760531eb9116e.)

In hardware, we force it to auto.  It doesn't have to be a fixed push of the mip bias, the game is already providing the slope to indicate that.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, that's true - implementing the slope should help - though hopefully the game will only use it on this detail texture and not the base texture, which will in such a case still be condemned to blurriness in the distance.
